cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -639,80 +574,109 @@ func (c *machineController) findNodeByProviderID(providerID normalizedProviderID
+		if err != nil {
+			return nil, err
+		}
+
+		scalableResources = append(scalableResources, machineDeployments...)
 	}
-	return newMachineDeploymentFromUnstructured(u.DeepCopy()), nil
+	return scalableResources, nil
 }
 
-func (c *machineController) listMachines(namespace string, selector labels.Selector) ([]*Machine, error) {
-	objs, err := c.machineInformer.Lister().ByNamespace(namespace).List(selector)
-	if err != nil {
-		return nil, err
+func (c *machineController) listResources(lister cache.GenericLister) ([]*unstructured.Unstructured, error) {
+	if len(c.autoDiscoverySpecs) == 0 {
+		return listResources(lister.ByNamespace(metav1.NamespaceAll), "", labels.Everything())
 	}
 
-	var machines []*Machine
-
-	for _, x := range objs {
-		u := x.(*unstructured.Unstructured).DeepCopy()
-		if machine := newMachineFromUnstructured(u); machine != nil {
-			machines = append(machines, machine)
+	var results []*unstructured.Unstructured
+	for _, spec := range c.autoDiscoverySpecs {
+		resources, err := listResources(lister.ByNamespace(spec.namespace), spec.clusterName, spec.labelSelector)
+		if err != nil {
+			return nil, err
 		}
+		results = append(results, resources...)
 	}
 
-	return machines, nil
+	return results, nil
 }
 
-func (c *machineController) listMachineSets(namespace string, selector labels.Selector) ([]*MachineSet, error) {
-	objs, err := c.machineSetInformer.Lister().ByNamespace(namespace).List(selector)
+func listResources(lister cache.GenericNamespaceLister, clusterName string, selector labels.Selector) ([]*unstructured.Unstructured, error) {
+	objs, err := lister.List(selector)
 	if err != nil {
 		return nil, err
 	}
 
-	var machineSets []*MachineSet
-
+	results := make([]*unstructured.Unstructured, 0, len(objs))
 	for _, x := range objs {
-		u := x.(*unstructured.Unstructured).DeepCopy()
-		if machineSet := newMachineSetFromUnstructured(u); machineSet != nil {
-			machineSets = append(machineSets, machineSet)
+		u, ok := x.(*unstructured.Unstructured)
+		if !ok {
+			return nil, fmt.Errorf("expected unstructured resource from lister, not %T", x)
 		}
+
+		// if clusterName is not empty and the clusterName does not match the resource, do not return it as part of the results
+		if clusterName != "" && clusterNameFromResource(u) != clusterName {
+			continue
+		}
+
+		// if we are listing MachineSets, do not return MachineSets that are owned by a MachineDeployment
+		if u.GetKind() == machineSetKind && machineSetHasMachineDeploymentOwnerRef(u) {
+			continue
+		}
+
+		results = append(results, u.DeepCopy())
 	}
 
-	return machineSets, nil
+	return results, nil
 }
 
-func (c *machineController) listMachineDeployments(namespace string, selector labels.Selector) ([]*MachineDeployment, error) {
-	objs, err := c.machineDeploymentInformer.Lister().ByNamespace(namespace).List(selector)
-	if err != nil {
-		return nil, err
+func (c *machineController) allowedByAutoDiscoverySpecs(r *unstructured.Unstructured) bool {
```

Comment:
Need to add test cases for this method to verify that autodiscovery config is doing the right thing.